gematria/datasets/pipelines/compile_modules_lib.py

```diff
@@ -0,0 +1,156 @@
+# Copyright 2024 Google Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import apache_beam as beam
+from rules_python.python.runfiles import runfiles
+import os
+
+from absl import logging
+
+from collections.abc import Iterable
+from collections.abc import Callable
+
+import subprocess
+
+from gematria.datasets.python import extract_bbs_from_obj
+
+
+def _get_llvm_binary_path(tool_name: str) -> str:
+  runfiles_env = runfiles.Create(os.environ)
+  assert runfiles_env is not None
+  return runfiles_env.Rlocation('llvm-project/llvm/' + tool_name)
+
+
+class OptimizeModules(beam.DoFn):
+  """A Beam function that uses LLVM opt to optimize bitcode modules."""
+
+  def __init__(self, optimization_pass_lists: list[str]):
```

Comment:
Also, based on how `self.optimization_pass_lists` is used below in `self.process()`, I'd expect it to be `Sequence[Sequence[str]]` rather than just `Sequence[str]`.

Note that `str` is also a `Sequence[str]` (a sequence of characters of the original string), so any type checks around sequences of strings must be treated with extra care.